### ISSUE TITLE
Error message when using DIY importer and dataset

### ISSUE THREAD

**Issue by SebastianScherer88** [mozilla/DeepSpeech#801]
Hi there,

i've written and importer for a different dataset that creates a csv file of the same type that is created by, for example, the "imported_ted.py" script already provided. when i am trying to train a model on this dataset using this csv file, i get the following error messages, and the training stops (doesn't even start, actually, bc this is happening right at the start). I've checked that all audio segment files are in the locations specified by the csv. Any pointers?

Error messages:

E Not enough time for target transition sequence (required: 32, available: 10)0You can turn this error into a warning by using the flag ignore_longer_outputs_than_inputs                          [12/1866]
E        [[Node: tower_0/CTCLoss = CTCLoss[ctc_merge_repeated=true, ignore_longer_outputs_than_inputs=false, preprocess_collapse_repeated=false, _device="/job:localhost/replica:0/task:0/cpu:0"](tower_0/Re
shape_7/_3975, tower_0/ToInt64/_3977, tower_0/Gather, tower_0/Gather_DequeueMany:1)]]
E        [[Node: tower_0/CTCLoss/_3979 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1$
 tensor_name="edge_36280_tower_0/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
E 
E Caused by op u'tower_0/CTCLoss', defined at:
E   File "DeepSpeech.py", line 1729, in <module>
E     tf.app.run()
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
E     _sys.exit(main(_sys.argv[:1] + flags_passthrough))
E   File "DeepSpeech.py", line 1689, in main
E     train()
E   File "DeepSpeech.py", line 1449, in train
E     results_tuple, gradients, mean_edit_distance, loss = get_tower_results(model_feeder, optimizer)
E   File "DeepSpeech.py", line 589, in get_tower_results
E     calculate_mean_edit_distance_and_loss(model_feeder, i, no_dropout if optimizer is None else dropout_rates)
E   File "DeepSpeech.py", line 476, in calculate_mean_edit_distance_and_loss
E     total_loss = tf.nn.ctc_loss(labels=batch_y, inputs=logits, sequence_length=batch_seq_len)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/ctc_ops.py", line 152, in ctc_loss
E     ignore_longer_outputs_than_inputs=ignore_longer_outputs_than_inputs)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/gen_ctc_ops.py", line 170, in _ctc_loss
E     name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
E     op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
E     original_op=self._default_original_op, op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
E     self._traceback = _extract_stack()
E 
E InvalidArgumentError (see above for traceback): Not enough time for target transition sequence (required: 32, available: 10)0You can turn this error into a warning by using the flag ignore_longer_outpu$s_than_inputs
E        [[Node: tower_0/CTCLoss = CTCLoss[ctc_merge_repeated=true, ignore_longer_outputs_than_inputs=false, preprocess_collapse_repeated=false, _device="/job:localhost/replica:0/task:0/cpu:0"](tower_0/R$shape_7/_3975, tower_0/ToInt64/_3977, tower_0/Gather, tower_0/Gather_DequeueMany:1)]]
E        [[Node: tower_0/CTCLoss/_3979 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1$ tensor_name="edge_36280_tower_0/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
E 
E You must feed a value for placeholder tensor 'Queue_Selector' with dtype int32
E        [[Node: Queue_Selector = Placeholder[dtype=DT_INT32, shape=<unknown>, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
E        [[Node: tower_0/boolean_mask/Where/_7729 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_in$arnation=1, tensor_name="edge_36144_tower_0/boolean_mask/Where", tensor_type=DT_INT64, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
E 
E Caused by op u'Queue_Selector', defined at:
E   File "DeepSpeech.py", line 1729, in <module>
E     tf.app.run()
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
E     _sys.exit(main(_sys.argv[:1] + flags_passthrough))
E   File "DeepSpeech.py", line 1689, in main
E     train()
E   File "DeepSpeech.py", line 1437, in train
E     tower_feeder_count=len(available_devices))
E   File "/home/ubuntu/projects/DeepSpeech_new/util/feeding.py", line 42, in __init__
E     self.ph_queue_selector = tf.placeholder(tf.int32, name='Queue_Selector')
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 1530, in placeholder
E     return gen_array_ops._placeholder(dtype=dtype, shape=shape, name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1954, in _placeholder
E     name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
E     op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
E     original_op=self._default_original_op, op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
E     self._traceback = _extract_stack()
E 
E InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'Queue_Selector' with dtype int32
E        [[Node: Queue_Selector = Placeholder[dtype=DT_INT32, shape=<unknown>, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
E        [[Node: tower_0/boolean_mask/Where/_7729 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_inc
arnation=1, tensor_name="edge_36144_tower_0/boolean_mask/Where", tensor_type=DT_INT64, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
E 
E Provide a --checkpoint_dir argument to work with models of different shapes.
2017-08-31 11:23:00.814971: F tensorflow/stream_executor/cuda/cuda_driver.cc:312] Check failed: CUDA_SUCCESS == cuCtxSetCurrent(cuda_context->context()) (0 vs. 4)

**Comment by reuben** [comment]
What's happening is that for some 〈audio file, transcription〉 pair in your dataset, the feature array that gets computed from the audio file is too short, and has less frames than the transcription has characters. In your case this seems particularly extreme, where the transcription has 32 characters and the feature array only has 10 frames. There are two possibilities:

1. The transcription is incorrect, and doesn't match the audio file. See for example [the Fisher importer](https://github.com/mozilla/DeepSpeech/blob/master/bin/import_fisher.py#L41). You can either remove them or fix the transcription.
2. The transcription is correct, but the audio file is tiny, usually something like "huh" or "hi". You have two options:
    1. If these are most of your dataset, remove the striding (done in util/audio.py and native_client/deepspeech_utils.cc), and possibly reduce the window step parameter used for the MFCCs.
    2. If these only happen a couple of times, throw them away. (See for example #790)

Hope this helps.

**Comment by reuben** [comment]
As for how to actually find the problematic files, I've used mostly printf-debugging in the past. Disabling the multi-GPU code and setting the number of importer threads to 1 really helps. Maybe tfdbg would be useful here, but I haven't used it.

**Comment by SebastianScherer88** [comment]
thanks for the quick answer! based on the #790 example, i think i'll start throwing away samples that include audios < ~3 seconds and see where i go from there. will come back if i hit a snag, but thanks for now!

**Comment by SebastianScherer88** [comment]
ok, i created a csv similar to the ldc93s1 set: one audio (~ 5 seconds), correct transcript (checked manually), and filesize and fed it to the model, and its still giving me the below error. it can't be missalignment, ive checked the audio and the transcript is good. in other words, none of the above two options seem to apply. i'm using the same parameters for training as for the ldc93s1 training (except stuff like data set path etc., obviously). bit stumped tbh. any ideas?

Error message:
E All labels must be nonnegative integers, batch: 0 labels: 6,15,18,0,8,5,18,0,12,1,19,20,0,-47,-43,0,25,5,1,18,19,0,20,8,15,21,7,8,0,19,8,5,0,3,15,21,12,4,0,23,1,12,11,0,19,8,5,0,4,9,4,0,14,15,20,0,12,5,
1,22,5,0,9,20
E        [[Node: tower_9/CTCLoss = CTCLoss[ctc_merge_repeated=true, ignore_longer_outputs_than_inputs=false, preprocess_collapse_repeated=false, _device="/job:localhost/replica:0/task:0/cpu:0"](tower_9/Re
shape_7/_2123, tower_9/ToInt64/_2125, tower_9/Gather, tower_9/Gather_9_DequeueMany:1)]]
E        [[Node: tower_9/CTCLoss/_2127 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:9", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1,
 tensor_name="edge_17080_tower_9/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:9"]()]]
E 
E Caused by op u'tower_9/CTCLoss', defined at:
E   File "DeepSpeech.py", line 1729, in <module>
E     tf.app.run()
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
E     _sys.exit(main(_sys.argv[:1] + flags_passthrough))
E   File "DeepSpeech.py", line 1689, in main
E     train()
E   File "DeepSpeech.py", line 1449, in train
E     results_tuple, gradients, mean_edit_distance, loss = get_tower_results(model_feeder, optimizer)
E   File "DeepSpeech.py", line 589, in get_tower_results
E     calculate_mean_edit_distance_and_loss(model_feeder, i, no_dropout if optimizer is None else dropout_rates)
E   File "DeepSpeech.py", line 476, in calculate_mean_edit_distance_and_loss
E     total_loss = tf.nn.ctc_loss(labels=batch_y, inputs=logits, sequence_length=batch_seq_len)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/ctc_ops.py", line 152, in ctc_loss
E     ignore_longer_outputs_than_inputs=ignore_longer_outputs_than_inputs)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/gen_ctc_ops.py", line 170, in _ctc_loss
E     name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
E     op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
E     original_op=self._default_original_op, op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
E     self._traceback = _extract_stack()
E 
E InvalidArgumentError (see above for traceback): All labels must be nonnegative integers, batch: 0 labels: 6,15,18,0,8,5,18,0,12,1,19,20,0,-47,-43,0,25,5,1,18,19,0,20,8,15,21,7,8,0,19,8,5,0,3,15,21,12,4,0,23,1,12,11,0,19,8,5,0,4,9,4,0,14,15,20,0,12,5,1,22,5,0,9,20
E        [[Node: tower_9/CTCLoss = CTCLoss[ctc_merge_repeated=true, ignore_longer_outputs_than_inputs=false, preprocess_collapse_repeated=false, _device="/job:localhost/replica:0/task:0/cpu:0"](tower_9/Reshape_7/_2123, tower_9/ToInt64/_2125, tower_9/Gather, tower_9/Gather_9_DequeueMany:1)]]
E        [[Node: tower_9/CTCLoss/_2127 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:9", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_17080_tower_9/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:9"]()]]
E 
E You must feed a value for placeholder tensor 'Queue_Selector' with dtype int32
E        [[Node: Queue_Selector = Placeholder[dtype=DT_INT32, shape=<unknown>, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
E        [[Node: tower_13/CTCLoss/_5121 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:13", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_8661_tower_13/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:13"]()]]
E 
E Caused by op u'Queue_Selector', defined at:
E   File "DeepSpeech.py", line 1729, in <module>
E     tf.app.run()
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
E     _sys.exit(main(_sys.argv[:1] + flags_passthrough))
E   File "DeepSpeech.py", line 1689, in main
E     train()
E   File "DeepSpeech.py", line 1437, in train
E     tower_feeder_count=len(available_devices))
E   File "/home/ubuntu/projects/DeepSpeech_new/util/feeding.py", line 42, in __init__
E     self.ph_queue_selector = tf.placeholder(tf.int32, name='Queue_Selector')
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 1530, in placeholder
E     return gen_array_ops._placeholder(dtype=dtype, shape=shape, name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1954, in _placeholder
E     name=name)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
E     op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
E     original_op=self._default_original_op, op_def=op_def)
E   File "/home/ubuntu/.local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
E     self._traceback = _extract_stack()
E 
E InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'Queue_Selector' with dtype int32
E        [[Node: Queue_Selector = Placeholder[dtype=DT_INT32, shape=<unknown>, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
E        [[Node: tower_13/CTCLoss/_5121 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:13", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_8661_tower_13/CTCLoss", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:13"]()]]
E 
E Provide a --checkpoint_dir argument to work with models of different shapes.

**Comment by reuben** [comment]
The problem here is different: you have invalid characters in the CSV file. In this case, numbers. You should probably normalize numbers to the corresponding words spoken, so change your text to be "'for her last *fifteen* years though she could walk she did not leave it".

**Comment by SebastianScherer88** [comment]
Yes, that did it! Thanks! This means that the models class output vocab (=labels of character softmax including 'blank') is preset, rather than adapting to the data being input, right? is there any way to access/edit that fixed set of characters that defines&labels the output layer (size) of the model, i.e. is there a script in 'utils' or something like that that can be easily edited? i guess this question relates to model adaption to other languages, too.

**Comment by reuben** [comment]
You're in luck, the code to handle that landed on master yesterday :)

See #797: you need to change `data/alphabet.txt`.

**Comment by reuben** [comment]
I'm closing this issue as it appears your original problem was fixed. Feel free to reopen if it was not.

**Comment by SebastianScherer88** [comment]
I have a question regarding the conversion of audio to array that is related to the"'tensorflow's ctc cant handle audios that have fewer frames than the target annotation has characters" issues: What is the conversion rate between seconds and audio frames, i.e. how many audio frames do i get out of, say, 1 second of audio; in other words, how finely are the audio chopped up when converted into arrays? Im asking because due to some missmatches in my diy importer, i seem to get short audios with long (incorrect) annotations. tensorflow says something like: "length of audio (100) is shorter than length of annotation (143)."

 now i know that all my audios are at least 2 seconds long (this i have verified). If i knew how many frames (-> "length of audio" from tensorflow's point of view) each second of audio produces, i could catch those audios during preprocessing with simple if "n_character > r_conv * n_secs = n_frames" for each audio segment. is there a fixed conversion rate r_conv? i have looked in the code but i havent been able to find that rate if it is there. 

many thanks in advance,

seb
